components/mpas-ocean/driver/ECDA_mpaso_obs_mod.F

```diff
@@ -0,0 +1,636 @@
+      if(present(obs_name)) OCN_obs_data(ID)%obs_name    = obs_name
+      if(present(filename)) OCN_obs_data(ID)%filename = filename
+      if(present(dm_name)) OCN_obs_data(ID)%dm_name = dm_name
+      if(present(inst_type)) OCN_obs_data(ID)%inst_type = inst_type
+      if(present(assimilate)) OCN_obs_data(ID)%assimilate = assimilate
+      if(present(required_states)) OCN_obs_data(ID)%required_states = required_states
+      if(present(update_states)) OCN_obs_data(ID)%update_states = update_states
+
+    end subroutine set_obs_type_value
+    
+!==============================================================================
+
+!BOP
+!   !ROUTINE: ocn_obs_type_register
+!   register the ocean observation type  for DA
+!
+!   !INTERFACE:
+
+  subroutine ocn_obs_type_register()
+!EOP
+!BOC
+        implicit none
+        ! --- Local variables ---
+        character(*), parameter :: subName =   '(ocn_obs_type_register) '
+        character(len = filename_len) :: nml_inpfile = 'ocean_da.namelist'
+        integer :: i,j,k,iob,iunit,rc,ierr
+        logical :: exists
+        logical :: obs_avalaible(max_OCN_obs_type)=.False.
+        character(len = type_name_len) ::  assim_obs_types(max_OCN_obs_type) = 'null'
+        namelist /obs_kind_nml/assim_obs_types
+        
+        NUM_OCN_obs_type=0
+        ! -read namelist for observation type        
+        inquire(file=trim(nml_inpfile), exist=exists)
+        if (.not. exists) return
+        iunit = shr_file_getUnit()
+        open(unit = iunit, file = trim(nml_inpfile), action = 'read')
+        read(unit = iunit, nml = obs_kind_nml, iostat = ierr)
+        close(unit = iunit)
+        call shr_file_freeUnit( iunit )
+        !- loop to get total number of assim. obs type
+        do i = 1, max_OCN_obs_type      
+           if(assim_obs_types(i) == 'null' .or. len_trim(assim_obs_types(i)) == 0 ) exit
+           NUM_OCN_obs_type = i
+        end do
+        if(NUM_OCN_obs_type == 0) return
+
+        !- initialize the observation variabel
+        allocate(OCN_obs_data(NUM_OCN_obs_type),stat=ierr)
+        call ocn_obs_init()
+
+        !- specific obs type  to assim, read the coresponding namelist file for
+        !assim. setting
+        do i= 1, NUM_OCN_obs_type  
+              OCN_obs_data(i)%obs_name=assim_obs_types(i)   
+              OCN_obs_data(i)%assimilate= .true.
+              call read_obs_assim_namelist(i,exists)   ! if ierr is not true, error, need revisit @ YL
+               RR=max(RR,OCN_obs_data(i)%RR_xy)         ! find maxium localization cutoff for  grid setting
+                ! active related state variable for DA
+              do k=1,n_OCN_state
+                 if( OCN_obs_data(i)%required_states(k).or. OCN_obs_data(i)%update_states(k)) &
+                     OCN_DAstate_data(k)%required = .true.
+              end do
+        end do
+        ! Falling off the end is an error
+         write(logunit,*) trim(subname), 'finish','NUM_OCN_obs_type',NUM_OCN_obs_type
```

Comment:
note that mpas has a specific logging capability it includes warnings and errors.  I think it would be good to switch to that capability (MPAS_LOG_WRITE)

Reply:
Sounds a good idea.